File: hash/js/rosetta-image-hash/src/phash.ts
import { Hash, ImageHashError, type RgbImage } from "./hash.js";
import { rgbToGray } from "./averageHash.js";
import { dct2d } from "./internal/dct.js";
import { toRgb, validateRgbImage } from "./internal/imgRgb.js";
import { resize } from "./internal/lanczos.js";

/**
 * ε threshold for the snap-to-threshold tie-break used by `phash`,
 * `phashSimple`, `whashDb4`, and `whashDb4Robust`. Coefficients within
 * `SNAP_EPS` of the threshold map deterministically to bit 0 across all
 * ports. Fixed across all 6 ports. See spec/SPEC.md §"Threshold tie-break".
 */
export const SNAP_EPS = 1e-10;

/**
 * phash: grayscale → Lanczos to (N*F, N*F) → 2-D DCT → top-left NxN block
 * → bit = (coefficient > median + {@link SNAP_EPS}).
 *
 * The snap-to-threshold tie-break (ε = 1e-10) deterministically maps
 * coefficients within ε of the median to bit 0, eliminating cross-port
 * FP-noise divergence at large hash sizes.
 */
export function phash(
  img: RgbImage,
  hashSize: number,
  highfreqFactor: number = 4,
): Hash {
  validateRgbImage(img);
  if (!Number.isInteger(hashSize)) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be an integer, got ${hashSize}`);
  }
  if (hashSize < 2) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be >= 2, got ${hashSize}`);
  }
  const imgSize = hashSize * highfreqFactor;

  const rgb = toRgb(img);
  const gray = rgbToGray(rgb.data, rgb.width, rgb.height);
  const resized = resize(gray, rgb.width, rgb.height, imgSize, imgSize);

  const doubles = new Float64Array(imgSize * imgSize);
  for (let i = 0; i < resized.length; i++) doubles[i] = resized[i];

  const dctOut = dct2d(doubles, imgSize);

  // Extract top-left hashSize×hashSize block
  const block = new Float64Array(hashSize * hashSize);
  let k = 0;
  for (let y = 0; y < hashSize; y++) {
    for (let x = 0; x < hashSize; x++) {
      block[k++] = dctOut[y * imgSize + x];
    }
  }
  const sorted = block.slice().sort();
  const n = sorted.length;
  const median = n % 2 === 1 ? sorted[(n - 1) / 2] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2;

  // Snap-to-threshold tie-break: deterministic bit 0 on ties.
  const threshold = median + SNAP_EPS;
  const bits: boolean[][] = [];
  for (let y = 0; y < hashSize; y++) {
    const row: boolean[] = new Array(hashSize);
    for (let x = 0; x < hashSize; x++) {
      row[x] = dctOut[y * imgSize + x] > threshold;
    }
    bits.push(row);
  }
  return new Hash(bits);
}

File: hash/rust/rosetta-image-hash/src/phash.rs
//! phash: grayscale → Lanczos resize to (N*F, N*F) → 2-D DCT → top-left NxN → median threshold.

use image::DynamicImage;

use crate::average::rgb_to_gray;
use crate::internal::{dct, img_rgb, lanczos};
use crate::{Hash, ImageHashError};

pub fn phash(img: &DynamicImage, hash_size: usize) -> Result<Hash, ImageHashError> {
    phash_with_factor(img, hash_size, 4)
}

pub fn phash_with_factor(
    img: &DynamicImage,
    hash_size: usize,
    highfreq_factor: usize,
) -> Result<Hash, ImageHashError> {
    if hash_size < 2 {
        return Err(ImageHashError::InvalidHashSize(hash_size));
    }
    let img_size = hash_size * highfreq_factor;

    let rgb = img_rgb::to_rgb(img);
    let gray = rgb_to_gray(&rgb);
    let resized = lanczos::resize(&gray, img_size, img_size);

    let mut doubles: Vec<Vec<f64>> = Vec::with_capacity(img_size);
    for y in 0..img_size {
        let mut row = Vec::with_capacity(img_size);
        for x in 0..img_size {
            row.push(f64::from(resized[y][x]));
        }
        doubles.push(row);
    }
    let dct_out = dct::dct2d(&doubles);

    let mut block: Vec<f64> = Vec::with_capacity(hash_size * hash_size);
    for y in 0..hash_size {
        for x in 0..hash_size {
            block.push(dct_out[y][x]);
        }
    }
    let mut sorted = block.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let n = sorted.len();
    let median = if n % 2 == 1 {
        sorted[n / 2]
    } else {
        (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0
    };

    let mut bits: Vec<Vec<bool>> = Vec::with_capacity(hash_size);
    for y in 0..hash_size {
        let mut row = Vec::with_capacity(hash_size);
        for x in 0..hash_size {
            row.push(dct_out[y][x] > median);
        }
        bits.push(row);
    }
    Ok(Hash::from_bits_unchecked(bits))
}
